src/main/java/de/tum/in/www1/artemis/service/InternalUrlService.java

```diff
@@ -0,0 +1,108 @@
+package de.tum.in.www1.artemis.service;
+
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Optional;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import de.tum.in.www1.artemis.domain.VcsRepositoryUrl;
+
+@Profile("bamboo")
+@Service
+public abstract class InternalUrlService {
+
+    protected Optional<URL> internalCiUrl;
+
+    protected Optional<URL> internalVcsUrl;
+
+    private static final Logger log = LoggerFactory.getLogger(InternalUrlService.class);
+
+    public InternalUrlService(Optional<URL> internalCiUrl, Optional<URL> internalVcsUrl) {
+        this.internalCiUrl = internalCiUrl;
+        this.internalVcsUrl = internalVcsUrl;
+    }
+
+    /**
+     * Replaces the url of the vcs repository url to the internal url if it's
+     * defined.
+     *
+     * @param vcsRepositoryUrl the vcs repository url
+     * @return the vcs repository url with the internal url
+     */
+    public VcsRepositoryUrl toInternalVcsUrl(VcsRepositoryUrl vcsRepositoryUrl) {
+        if (vcsRepositoryUrl.getURI() == null) {
+            log.warn("Cannot replace url to internal url {} because the url is null.", internalVcsUrl);
+            return vcsRepositoryUrl;
+        }
+
+        try {
+            String newInternalUrl = toInternalVcsUrl(vcsRepositoryUrl.getURI().toString());
+            return new VcsRepositoryUrl(newInternalUrl);
+        }
+        catch (URISyntaxException e) {
+            log.warn("Cannot replace url {} to {}: {}. Falling back to original url.", vcsRepositoryUrl, internalVcsUrl, e.getMessage());
```

Comment:
When can this ever happen? Is just using the old URI a good fallback? Could it happen that we end up with URLs of both types in the end?

Reply:
It can happen if either the configuration is wrong and the internal host is not valid, we then simply return the old URL, as there is nothing we can do about it. This code is not new, I simply moved it to a new class to make use of internal URLs in Bamboo, Jenkins uses that class since 2021.